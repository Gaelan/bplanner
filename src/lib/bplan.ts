export interface Load {
	type: string;
	trailingLoad: string;
	speed: string;
	raGauge: string;
	description: string;
	itpsPowerType: string;
	itpsLoad: string;
	itpsSpeed: string;
}

export interface Location {
	tiploc: string;
	name: string;
	northing: number;
	easting: number;
	timingPointType: 'trust' | 'mandatory' | 'optional';
	zone: string;
	stanox: number;
	offNetwork: boolean;
	forceLine: 'approaching' | 'leaving' | 'both' | 'none';
}

export interface Link {
	from: Location;
	to: Location;
	line: string;
	lineDesc: string;
	fromDir: 'up' | 'down';
	toDir: 'up' | 'down';
	distance: number;
	dooPassenger: boolean;
	dooNonPassenger: boolean;
	retb: boolean;
	zone: string;
	reversible: 'bidirectional' | 'reversible' | 'neither';
	powerSupply: string;
	ra: number;
	maxTrainLength: number;
	timings: Timing[];
}

export interface Timing {
	link: Link;
	load: Load;
	fromStart: boolean;
	toStop: boolean;
	time: number; //seconds
}

export interface Bplan {
	loads: Record<string, Load>; // key: loadKey()
	locations: Record<string, Location>; // key: tiploc
	links: Record<string, Record<string, Link[]>>; // keys: from and to tiploc
	timings: Timing[];
}

export function loadKey(load: Load) {
	return [load.type, load.trailingLoad, load.speed, load.raGauge].join(' ');
}

export async function loadBplan(): Promise<Bplan> {
	// it's named this to avoid sveltekit serving it as content-encoding: gz
	// in development, when we can't guarantee this happen in production
	const req = await fetch('/bplan.gz.bleh');
	if (!req.ok || !req.body) {
		throw new Error('failed to fetch bplan data');
	}

	return loadBplanFromCompressedStream(req.body);
}

export async function loadBplanFromCompressedStream(stream: ReadableStream): Promise<Bplan> {
	return parseStream(
		stream.pipeThrough(new DecompressionStream('gzip')).pipeThrough(new TextDecoderStream('utf-8'))
	);
}

function handleLine(bplan: Bplan, line: string) {
	const parts = line.split('\t');

	if (parts[0] == 'TLD') {
		const load: Load = {
			// 0 "TLD"
			// 1 "A"
			type: parts[2].trim(),
			trailingLoad: parts[3].trim(),
			speed: parts[4].trim(),
			raGauge: parts[5].trim(),
			description: parts[6],
			itpsPowerType: parts[7],
			itpsLoad: parts[8],
			itpsSpeed: parts[9]
		};
		bplan.loads[loadKey(load)] = load;
	} else if (parts[0] == 'LOC') {
		if (parts[5] != '') {
			console.log(parts);
		}
		const loc = {
			// 0 "LOC"
			// 1 "A"
			tiploc: parts[2],
			name: parts[3],
			// 4 start date
			// 5 end date
			easting: parseInt(parts[6]),
			northing: parseInt(parts[7]),
			timingPointType: ({ T: 'trust', M: 'mandatory', O: 'optional' } as const)[parts[8]]!,
			zone: parts[9],
			stanox: parseInt(parts[10]),
			offNetwork: parts[11] == 'Y',
			forceLine: ({ L: 'approaching', P: 'leaving', B: 'both', ' ': 'none' } as const)[parts[12]]!
		};
		bplan.locations[loc.tiploc] = loc;
	} else if (parts[0] == 'NWK') {
		if (parts[7] != '') {
			// console.log('expires', parts);
		}
		const link: Link = {
			// 0 "NWK"
			// 1 "A"
			from: bplan.locations[parts[2]],
			to: bplan.locations[parts[3]],
			line: parts[4].trim(),
			lineDesc: parts[5],
			// 6 start date
			// 7 end date
			fromDir: ({ U: 'up', D: 'down' } as const)[parts[8]]!,
			toDir: ({ U: 'up', D: 'down' } as const)[parts[9]]!,
			distance: parseInt(parts[10]),
			dooPassenger: parts[11] == 'Y',
			dooNonPassenger: parts[12] == 'Y',
			retb: parts[13] == 'Y',
			zone: parts[14],
			reversible: ({ R: 'reversible', B: 'bidirectional', N: 'neither' } as const)[parts[15]]!,
			powerSupply: parts[16],
			ra: parseInt(parts[17]),
			maxTrainLength: parseInt(parts[18]),
			timings: []
		};
		if (!link.from) {
			console.log(parts);
			return;
		}
		bplan.links[link.from.tiploc] ||= {};
		bplan.links[link.from.tiploc][link.to.tiploc] ||= [];
		bplan.links[link.from.tiploc][link.to.tiploc].push(link);
	} else if (parts[0] == 'TLK') {
		if (parts[7] != '') {
			// console.log(parts);
		}

		const timeParts = parts[13].split("'");
		const timing = {
			// 0 "TLK"
			// 1 "A"
			link: bplan.links[parts[2]][parts[3]].find((l) => l.line == parts[4].trim())!,
			load: bplan.loads[
				parts
					.slice(5, 9)
					.map((x) => x.trim())
					.join(' ')
			],
			fromStart: parts[9] == '0',
			toStop: parts[10] == '0',
			// 11 start date
			// 12 end date
			time: parseInt(timeParts[0]) * 60 + parseInt(timeParts[1]),
			description: parts[14]
		};

		timing.link.timings.push(timing);
		bplan.timings.push(timing);
	}
}

async function parseStream(stream: ReadableStream<string>): Promise<Bplan> {
	const bplan = {
		locations: {},
		links: {},
		loads: {},
		timings: []
	};

	let currentLine = '';

	const reader = stream.getReader();

	while (true) {
		const result = await reader.read();
		if (result.done) {
			break;
		}

		const lineParts = result.value.split('\n');

		if (lineParts.length == 1) {
			currentLine += lineParts[0];
		} else {
			currentLine += lineParts[0];
			handleLine(bplan, currentLine);
			lineParts.slice(1, -1).forEach((x) => handleLine(bplan, x));
			currentLine = lineParts[lineParts.length - 1];
		}
	}

	return bplan;
}
