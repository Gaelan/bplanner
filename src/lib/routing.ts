import { loadKey, type Bplan, type Link, type Timing } from './bplan';
import { MinPriorityQueue, PriorityQueue } from '@datastructures-js/priority-queue';

export function route(
	bplan: Bplan,
	fromTiploc: string,
	toTiploc: string,
	banLoads: string[]
): [number, Timing[]] {
	let fromNode = 's' + fromTiploc;
	let toNode = 's' + toTiploc;
	let prev: Record<string, Timing> = {};
	let dists = { [fromNode]: 0 };
	let done = new Set();
	let queue = new MinPriorityQueue<[number, string]>((x) => x[0]);

	queue.enqueue([0, fromNode]);

	outer: while (queue.size() > 0) {
		const current = queue.dequeue()[1];
		const currentTiploc = current.slice(1);
		const currentIsStop = current[0] == 's';
		if (done.has(current)) {
			continue;
		}
		if (current == toNode) {
			break outer;
		}

		for (const linkTo in bplan.links[currentTiploc]) {
			for (const link of bplan.links[currentTiploc][linkTo]) {
				for (const timing of link.timings) {
					if (timing.fromStart != currentIsStop) {
						continue;
					}
					if (banLoads.includes(loadKey(timing.load))) {
						continue;
					}
					const nextNode = (timing.toStop ? 's' : 'f') + linkTo;
					const alt = dists[current] + timing.time;
					if (!dists[nextNode] || alt < dists[nextNode]) {
						prev[nextNode] = timing;
						dists[nextNode] = alt;

						// we can't decrease priority; instead we allow duplicate
						// entries in the queue and filter them out with done
						queue.enqueue([alt, nextNode]);
					}
				}
			}
		}

		done.add(current);
	}

	if (!dists[toNode]) {
		throw new Error('pathfinding failed');
	}

	// trace back
	let hist = [prev[toNode]];
	while (true) {
		const prevNodeName =
			(hist[hist.length - 1].fromStart ? 's' : 'f') + hist[hist.length - 1].link.from.tiploc;
		if (prevNodeName == fromNode) {
			break;
		}
		hist.push(prev[prevNodeName]);
	}

	// hist = simplifyLoads(hist);

	hist = simplifyLoads(hist);

	hist.reverse();

	hist = simplifyLoads(hist);

	return [dists[toNode], hist];
}

function simplifyLoads(hist: Timing[]) {
	const seenLoads = new Set();

	return hist.map((timing) => {
		const betterTiming = timing.link.timings.find(
			(t) =>
				t.fromStart == timing.fromStart &&
				t.toStop == timing.toStop &&
				t.time <= timing.time &&
				seenLoads.has(loadKey(t.load))
		);

		const ret = betterTiming || timing;

		seenLoads.add(loadKey(ret.load));

		return ret;
	});
}
