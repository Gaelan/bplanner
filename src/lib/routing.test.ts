import { describe, it, expect, beforeAll } from 'vitest';
import { route } from './routing';
import { loadBplan, loadBplanFromCompressedStream, loadKey, type Bplan } from './bplan';
import { createReadableStreamFromReadable } from './testUtil';
import { createReadStream } from 'node:fs';

let bplan: Bplan;

beforeAll(async () => {
	bplan = await loadBplanFromCompressedStream(
		createReadableStreamFromReadable(createReadStream('static/bplan.gz.bleh'))
	);
}); //.map((x) => [x.link.from.tiploc, x.link.to.tiploc, x.load.description, x.time])

describe('route', () => {
	it('Leuchars-Edinburgh', async () => {
		const [dist, hist] = route(bplan, 'LEUCHRS', true, 'EDINBUR', true, [], []);

		expect([
			dist,
			hist.map((x) => [x.link.from.tiploc, x.link.to.tiploc, x.load.description, x.time])
		]).toMatchSnapshot();
	});

	it('Leuchars-Edinburgh with preferred loads', async () => {
		const [dist, hist] = route(bplan, 'LEUCHRS', true, 'EDINBUR', true, ['170  100 '], []);

		expect([
			dist,
			hist.map((x) => [x.link.from.tiploc, x.link.to.tiploc, x.load.description, x.time])
		]).toMatchSnapshot();
	});

	it('Leuchars-London', async () => {
		const [dist, hist] = route(bplan, 'LEUCHRS', true, 'KNGX', true, [], []);

		expect([
			dist,
			hist.map((x) => [x.link.from.tiploc, x.link.to.tiploc, x.load.description, x.time])
		]).toMatchSnapshot();
	});

	it('Leuchars-London with banned loads', async () => {
		const [dist, hist] = route(bplan, 'LEUCHRS', true, 'KNGX', true, [], ['BUS   ']);

		expect([
			dist,
			hist.map((x) => [x.link.from.tiploc, x.link.to.tiploc, x.load.description, x.time])
		]).toMatchSnapshot();
	});
});
