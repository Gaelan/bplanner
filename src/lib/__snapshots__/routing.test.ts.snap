// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`route > Leuchars-Edinburgh 1`] = `
[
  2580,
  [
    [
      "LEUCHRS",
      "CUPAR",
      "Bombardier Non Tilt DMU",
      300,
    ],
    [
      "CUPAR",
      "LADYBNK",
      "Bombardier Non Tilt DMU",
      270,
    ],
    [
      "LADYBNK",
      "THRNTNJ",
      "HST 2+4 100mph Scotrail",
      360,
    ],
    [
      "THRNTNJ",
      "THRNTSJ",
      "Bombardier Non Tilt DMU",
      30,
    ],
    [
      "THRNTSJ",
      "KCLD",
      "Bombardier Non Tilt DMU",
      240,
    ],
    [
      "KCLD",
      "BISLND",
      "Class 47 & 10 coaches (95 mph)",
      180,
    ],
    [
      "BISLND",
      "IVRKTHG",
      "Bombardier Non Tilt DMU",
      420,
    ],
    [
      "IVRKTHG",
      "DLMYJN",
      "HST 2+4 100mph Scotrail",
      270,
    ],
    [
      "DLMYJN",
      "HAYMRWJ",
      "HST 2+4 100mph Scotrail",
      270,
    ],
    [
      "HAYMRWJ",
      "HAYMRKT",
      "Class 158",
      60,
    ],
    [
      "HAYMRKT",
      "PRNCSTG",
      "Class 158",
      90,
    ],
    [
      "PRNCSTG",
      "EDINBUR",
      "Class 800 9 Car Diesel",
      90,
    ],
  ],
]
`;

exports[`route > Leuchars-Edinburgh with preferred loads 1`] = `
[
  3150,
  [
    [
      "LEUCHRS",
      "CUPAR",
      "Class 170",
      390,
    ],
    [
      "CUPAR",
      "LADYBNK",
      "Class 170",
      300,
    ],
    [
      "LADYBNK",
      "THRNTNJ",
      "Class 170",
      450,
    ],
    [
      "THRNTNJ",
      "THRNTSJ",
      "Class 170",
      30,
    ],
    [
      "THRNTSJ",
      "KCLD",
      "Class 170",
      240,
    ],
    [
      "KCLD",
      "BISLND",
      "Class 170",
      390,
    ],
    [
      "BISLND",
      "IVRKTHG",
      "Class 170",
      450,
    ],
    [
      "IVRKTHG",
      "DLMYJN",
      "Class 170",
      330,
    ],
    [
      "DLMYJN",
      "HAYMRWJ",
      "Class 170",
      300,
    ],
    [
      "HAYMRWJ",
      "HAYMRKT",
      "Class 170",
      60,
    ],
    [
      "HAYMRKT",
      "PRNCSTG",
      "Class 170",
      90,
    ],
    [
      "PRNCSTG",
      "EDINBUR",
      "Class 170",
      120,
    ],
  ],
]
`;

exports[`route > Leuchars-London 1`] = `
[
  14310,
  [
    [
      "LEUCHRS",
      "CUPAR",
      "Bombardier Non Tilt DMU",
      300,
    ],
    [
      "CUPAR",
      "LADYBNK",
      "Bombardier Non Tilt DMU",
      270,
    ],
    [
      "LADYBNK",
      "THRNTNJ",
      "HST 2+4 100mph Scotrail",
      360,
    ],
    [
      "THRNTNJ",
      "THRNTSJ",
      "Bombardier Non Tilt DMU",
      30,
    ],
    [
      "THRNTSJ",
      "KCLD",
      "Bombardier Non Tilt DMU",
      240,
    ],
    [
      "KCLD",
      "BISLND",
      "Class 47 & 10 coaches (95 mph)",
      180,
    ],
    [
      "BISLND",
      "IVRKTHG",
      "Bombardier Non Tilt DMU",
      420,
    ],
    [
      "IVRKTHG",
      "DLMYJN",
      "HST 2+4 100mph Scotrail",
      270,
    ],
    [
      "DLMYJN",
      "HAYMRWJ",
      "HST 2+4 100mph Scotrail",
      270,
    ],
    [
      "HAYMRWJ",
      "HAYMRKT",
      "Class 800 9 Car Electric",
      60,
    ],
    [
      "HAYMRKT",
      "CRSTRS",
      "Bus",
      600,
    ],
    [
      "CRSTRS",
      "CRSTRSS",
      "Class 150, 155 or 156",
      60,
    ],
    [
      "CRSTRSS",
      "ABINGTN",
      "Class 390 EMU - 9 vehicles 125 mph",
      480,
    ],
    [
      "ABINGTN",
      "BEATCKS",
      "Class 67 & 4 coaches (125 mph)",
      270,
    ],
    [
      "BEATCKS",
      "BEATCK",
      "Class 390 EMU - 9 vehicles 125 mph",
      300,
    ],
    [
      "BEATCK",
      "LCKRBIE",
      "Class 390 EMU - 9 vehicles 125 mph",
      420,
    ],
    [
      "LCKRBIE",
      "KRKP862",
      "Class 390 EMU - 9 vehicles 125 mph",
      390,
    ],
    [
      "KRKP862",
      "QNTNSHL",
      "Class 86/87/90 & 10 coaches (100 mph)",
      60,
    ],
    [
      "QNTNSHL",
      "GRETNAJ",
      "Bombardier Non Tilt DMU",
      30,
    ],
    [
      "GRETNAJ",
      "CARLILE",
      "Class 390 EMU - 9 vehicles 125 mph",
      360,
    ],
    [
      "CARLILE",
      "PNTH",
      "Class 90 Freight 500 tonnes at 50mph",
      150,
    ],
    [
      "PNTH",
      "SHAPSMT",
      "Class 325 Electric Parcels Unit",
      480,
    ],
    [
      "SHAPSMT",
      "TEBAY",
      "Class 325 Electric Parcels Unit",
      240,
    ],
    [
      "TEBAY",
      "GRIGG",
      "Class 325 Electric Parcels Unit",
      240,
    ],
    [
      "GRIGG",
      "OXENHLM",
      "Class 390 EMU - 9 vehicles 125 mph",
      270,
    ],
    [
      "OXENHLM",
      "CRNFNJN",
      "Class 390 EMU - 9 vehicles 125 mph",
      420,
    ],
    [
      "CRNFNJN",
      "MORCMSJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      120,
    ],
    [
      "MORCMSJ",
      "LANCSTR",
      "Class 90 & 7 coaches (110 mph)",
      60,
    ],
    [
      "LANCSTR",
      "GSTANG",
      "Class 390 EMU - 9 vehicles 125 mph",
      360,
    ],
    [
      "GSTANG",
      "PRSTNFJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      300,
    ],
    [
      "PRSTNFJ",
      "PRST",
      "Class 195 2 car",
      60,
    ],
    [
      "PRST",
      "MNCRIAP",
      "Bus",
      60,
    ],
    [
      "MNCRIAP",
      "HLDGWJ",
      "Class 195 2 car",
      90,
    ],
    [
      "HLDGWJ",
      "HLDGSJ",
      "Class 150, 155 or 156",
      60,
    ],
    [
      "HLDGSJ",
      "WLMSL",
      "Class 92 & 16 coaches at 80mph (inc. comfort speeds)",
      120,
    ],
    [
      "WLMSL",
      "ALDEDGE",
      "Class 150, 155 or 156",
      60,
    ],
    [
      "ALDEDGE",
      "SBCH",
      "Class 350 110mph",
      420,
    ],
    [
      "SBCH",
      "CREW154",
      "Class 390 EMU - 9 vehicles 125 mph",
      180,
    ],
    [
      "CREW154",
      "CREWE",
      "Class 150, 155 or 156",
      120,
    ],
    [
      "CREWE",
      "CREWBHJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      60,
    ],
    [
      "CREWBHJ",
      "MADELEY",
      "Class 390 EMU - 9 vehicles 125 mph",
      210,
    ],
    [
      "MADELEY",
      "NTNB",
      "Class 390 EMU - 9 vehicles 125 mph",
      300,
    ],
    [
      "NTNB",
      "STAFFRD",
      "Class 390 EMU - 9 vehicles 125 mph",
      180,
    ],
    [
      "STAFFRD",
      "COLWICH",
      "Bombardier Non Tilt DMU",
      270,
    ],
    [
      "COLWICH",
      "RUGLYNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      90,
    ],
    [
      "RUGLYNJ",
      "LCHTNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      240,
    ],
    [
      "LCHTNJ",
      "AMNGTNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      210,
    ],
    [
      "AMNGTNJ",
      "NNTN",
      "Class 390 EMU - 9 vehicles 125 mph",
      360,
    ],
    [
      "NNTN",
      "RUGBTVJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      390,
    ],
    [
      "RUGBTVJ",
      "RUGBY",
      "Class 390 EMU - 9 vehicles 125 mph",
      30,
    ],
    [
      "RUGBY",
      "HMTNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      30,
    ],
    [
      "HMTNJ",
      "WEEDON",
      "Class 90 Freight 1000 tonnes at 50mph",
      90,
    ],
    [
      "WEEDON",
      "HANSLPJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      390,
    ],
    [
      "HANSLPJ",
      "MKNSCEN",
      "Class 390 EMU - 9 vehicles 125 mph",
      180,
    ],
    [
      "MKNSCEN",
      "BLTCHLY",
      "Bombardier Tilt DMU",
      90,
    ],
    [
      "BLTCHLY",
      "LEDBRNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      270,
    ],
    [
      "LEDBRNJ",
      "TRING",
      "Class 390 EMU - 9 vehicles 125 mph",
      150,
    ],
    [
      "TRING",
      "BONENDJ",
      "Bombardier Tilt DMU",
      180,
    ],
    [
      "BONENDJ",
      "WATFDJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      240,
    ],
    [
      "WATFDJ",
      "HROW",
      "Class 66 Freight - 2200 tonnes at 60 mph",
      120,
    ],
    [
      "HROW",
      "WMBY",
      "Bombardier Tilt DMU",
      120,
    ],
    [
      "WMBY",
      "WLSDWLJ",
      "Bombardier Tilt DMU",
      90,
    ],
    [
      "WLSDWLJ",
      "CMDNJN",
      "Bombardier Tilt DMU",
      180,
    ],
    [
      "CMDNJN",
      "CMDNRDJ",
      "Class 150, 155 or 156",
      180,
    ],
    [
      "CMDNRDJ",
      "CMDNRCJ",
      "Class 92 & 16 coaches at 80mph (inc. comfort speeds)",
      60,
    ],
    [
      "CMDNRCJ",
      "YWAYNJN",
      "Class 47 & 13 coaches (75 mph)",
      90,
    ],
    [
      "YWAYNJN",
      "KNGXFTJ",
      "Class 66 Freight - 2200 tonnes at 60 mph",
      120,
    ],
    [
      "KNGXFTJ",
      "KNGXBEL",
      "Class 180",
      90,
    ],
    [
      "KNGXBEL",
      "KNGX",
      "Class 180",
      120,
    ],
  ],
]
`;

exports[`route > Leuchars-London with banned loads 1`] = `
[
  14310,
  [
    [
      "LEUCHRS",
      "CUPAR",
      "Bombardier Non Tilt DMU",
      300,
    ],
    [
      "CUPAR",
      "LADYBNK",
      "Bombardier Non Tilt DMU",
      270,
    ],
    [
      "LADYBNK",
      "THRNTNJ",
      "HST 2+4 100mph Scotrail",
      360,
    ],
    [
      "THRNTNJ",
      "THRNTSJ",
      "Bombardier Non Tilt DMU",
      30,
    ],
    [
      "THRNTSJ",
      "KCLD",
      "Bombardier Non Tilt DMU",
      240,
    ],
    [
      "KCLD",
      "BISLND",
      "Class 47 & 10 coaches (95 mph)",
      180,
    ],
    [
      "BISLND",
      "IVRKTHG",
      "Bombardier Non Tilt DMU",
      420,
    ],
    [
      "IVRKTHG",
      "DLMYJN",
      "HST 2+4 100mph Scotrail",
      270,
    ],
    [
      "DLMYJN",
      "HAYMRWJ",
      "HST 2+4 100mph Scotrail",
      270,
    ],
    [
      "HAYMRWJ",
      "HAYMRKT",
      "Class 800 9 Car Electric",
      60,
    ],
    [
      "HAYMRKT",
      "CRSTRS",
      "Bus",
      600,
    ],
    [
      "CRSTRS",
      "CRSTRSS",
      "Class 150, 155 or 156",
      60,
    ],
    [
      "CRSTRSS",
      "ABINGTN",
      "Class 390 EMU - 9 vehicles 125 mph",
      480,
    ],
    [
      "ABINGTN",
      "BEATCKS",
      "Class 67 & 4 coaches (125 mph)",
      270,
    ],
    [
      "BEATCKS",
      "BEATCK",
      "Class 390 EMU - 9 vehicles 125 mph",
      300,
    ],
    [
      "BEATCK",
      "LCKRBIE",
      "Class 390 EMU - 9 vehicles 125 mph",
      420,
    ],
    [
      "LCKRBIE",
      "KRKP862",
      "Class 390 EMU - 9 vehicles 125 mph",
      390,
    ],
    [
      "KRKP862",
      "QNTNSHL",
      "Class 86/87/90 & 10 coaches (100 mph)",
      60,
    ],
    [
      "QNTNSHL",
      "GRETNAJ",
      "Bombardier Non Tilt DMU",
      30,
    ],
    [
      "GRETNAJ",
      "CARLILE",
      "Class 390 EMU - 9 vehicles 125 mph",
      360,
    ],
    [
      "CARLILE",
      "PNTH",
      "Class 90 Freight 500 tonnes at 50mph",
      150,
    ],
    [
      "PNTH",
      "SHAPSMT",
      "Class 325 Electric Parcels Unit",
      480,
    ],
    [
      "SHAPSMT",
      "TEBAY",
      "Class 325 Electric Parcels Unit",
      240,
    ],
    [
      "TEBAY",
      "GRIGG",
      "Class 325 Electric Parcels Unit",
      240,
    ],
    [
      "GRIGG",
      "OXENHLM",
      "Class 390 EMU - 9 vehicles 125 mph",
      270,
    ],
    [
      "OXENHLM",
      "CRNFNJN",
      "Class 390 EMU - 9 vehicles 125 mph",
      420,
    ],
    [
      "CRNFNJN",
      "MORCMSJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      120,
    ],
    [
      "MORCMSJ",
      "LANCSTR",
      "Class 90 & 7 coaches (110 mph)",
      60,
    ],
    [
      "LANCSTR",
      "GSTANG",
      "Class 390 EMU - 9 vehicles 125 mph",
      360,
    ],
    [
      "GSTANG",
      "PRSTNFJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      300,
    ],
    [
      "PRSTNFJ",
      "PRST",
      "Class 195 2 car",
      60,
    ],
    [
      "PRST",
      "MNCRIAP",
      "Bus",
      60,
    ],
    [
      "MNCRIAP",
      "HLDGWJ",
      "Class 195 2 car",
      90,
    ],
    [
      "HLDGWJ",
      "HLDGSJ",
      "Class 150, 155 or 156",
      60,
    ],
    [
      "HLDGSJ",
      "WLMSL",
      "Class 92 & 16 coaches at 80mph (inc. comfort speeds)",
      120,
    ],
    [
      "WLMSL",
      "ALDEDGE",
      "Class 150, 155 or 156",
      60,
    ],
    [
      "ALDEDGE",
      "SBCH",
      "Class 350 110mph",
      420,
    ],
    [
      "SBCH",
      "CREW154",
      "Class 390 EMU - 9 vehicles 125 mph",
      180,
    ],
    [
      "CREW154",
      "CREWE",
      "Class 150, 155 or 156",
      120,
    ],
    [
      "CREWE",
      "CREWBHJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      60,
    ],
    [
      "CREWBHJ",
      "MADELEY",
      "Class 390 EMU - 9 vehicles 125 mph",
      210,
    ],
    [
      "MADELEY",
      "NTNB",
      "Class 390 EMU - 9 vehicles 125 mph",
      300,
    ],
    [
      "NTNB",
      "STAFFRD",
      "Class 390 EMU - 9 vehicles 125 mph",
      180,
    ],
    [
      "STAFFRD",
      "COLWICH",
      "Bombardier Non Tilt DMU",
      270,
    ],
    [
      "COLWICH",
      "RUGLYNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      90,
    ],
    [
      "RUGLYNJ",
      "LCHTNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      240,
    ],
    [
      "LCHTNJ",
      "AMNGTNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      210,
    ],
    [
      "AMNGTNJ",
      "NNTN",
      "Class 390 EMU - 9 vehicles 125 mph",
      360,
    ],
    [
      "NNTN",
      "RUGBTVJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      390,
    ],
    [
      "RUGBTVJ",
      "RUGBY",
      "Class 390 EMU - 9 vehicles 125 mph",
      30,
    ],
    [
      "RUGBY",
      "HMTNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      30,
    ],
    [
      "HMTNJ",
      "WEEDON",
      "Class 90 Freight 1000 tonnes at 50mph",
      90,
    ],
    [
      "WEEDON",
      "HANSLPJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      390,
    ],
    [
      "HANSLPJ",
      "MKNSCEN",
      "Class 390 EMU - 9 vehicles 125 mph",
      180,
    ],
    [
      "MKNSCEN",
      "BLTCHLY",
      "Bombardier Tilt DMU",
      90,
    ],
    [
      "BLTCHLY",
      "LEDBRNJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      270,
    ],
    [
      "LEDBRNJ",
      "TRING",
      "Class 390 EMU - 9 vehicles 125 mph",
      150,
    ],
    [
      "TRING",
      "BONENDJ",
      "Bombardier Tilt DMU",
      180,
    ],
    [
      "BONENDJ",
      "WATFDJ",
      "Class 390 EMU - 9 vehicles 125 mph",
      240,
    ],
    [
      "WATFDJ",
      "HROW",
      "Class 66 Freight - 2200 tonnes at 60 mph",
      120,
    ],
    [
      "HROW",
      "WMBY",
      "Bombardier Tilt DMU",
      120,
    ],
    [
      "WMBY",
      "WLSDWLJ",
      "Bombardier Tilt DMU",
      90,
    ],
    [
      "WLSDWLJ",
      "CMDNJN",
      "Bombardier Tilt DMU",
      180,
    ],
    [
      "CMDNJN",
      "CMDNRDJ",
      "Class 150, 155 or 156",
      180,
    ],
    [
      "CMDNRDJ",
      "CMDNRCJ",
      "Class 92 & 16 coaches at 80mph (inc. comfort speeds)",
      60,
    ],
    [
      "CMDNRCJ",
      "YWAYNJN",
      "Class 47 & 13 coaches (75 mph)",
      90,
    ],
    [
      "YWAYNJN",
      "KNGXFTJ",
      "Class 66 Freight - 2200 tonnes at 60 mph",
      120,
    ],
    [
      "KNGXFTJ",
      "KNGXBEL",
      "Class 180",
      90,
    ],
    [
      "KNGXBEL",
      "KNGX",
      "Class 180",
      120,
    ],
  ],
]
`;
